tikv/raftstore/worker.go

```diff
@@ -290,7 +291,75 @@ type pendingDeleteRanges struct {
 }
 
 func (pendDelRanges *pendingDeleteRanges) insert(regionId uint64, startKey, endKey []byte, timeout time.Time) {
-	// todo, currently it is a place holder.
+	if len(pendDelRanges.findOverlapRanges(startKey, endKey)) != 0 {
+		panic(fmt.Sprintf("[region %d] register deleting data in [%v, %v) failed due to overlap", regionId, startKey, endKey))
+	}
+	peerInfo := &stalePeerInfo{
+		endKey:   endKey,
+		regionId: regionId,
+		timeout:  timeout,
+	}
+	pendDelRanges.ranges.Insert(startKey, peerInfo.serializeStalePeerInfo())
+}
+
+// remove removes and returns the peer info with the `start_key`.
+func (pendDelRanges *pendingDeleteRanges) remove(startKey []byte) *stalePeerInfo {
+	value := pendDelRanges.ranges.Get(startKey, nil)
+	if value != nil {
+		pendDelRanges.ranges.Delete(startKey)
+		return deserializeStalePeerInfo(safeCopy(value))
+	}
+	return nil
+}
+
+// timeoutRanges returns all timeout ranges info.
+func (pendDelRanges *pendingDeleteRanges) timeoutRanges(now time.Time) (ranges []delRangeHolder) {
+	ite := pendDelRanges.ranges.NewIterator()
+	for ite.Next(); ite.Valid(); ite.Next() {
+		startKey := safeCopy(ite.Key())
+		peerInfo := deserializeStalePeerInfo(safeCopy(ite.Value()))
+		if peerInfo.timeout.Before(now) {
+			ranges = append(ranges, delRangeHolder{
+				startKey: startKey,
+				endKey:   peerInfo.endKey,
+				regionId: peerInfo.regionId,
+			})
+		}
+	}
+	return
+}
+
+type stalePeerInfo struct {
```

Comment:
You can define this type like this
```
type stalePeerInfo struct {
    data []byte
}
```
Then define methods to get those fields.
So we don't need to call serialize and deserialize.